dbrp/service.go

```diff
@@ -0,0 +1,490 @@
+		return err
+	}
+
+	// If a dbrp with this particular ID already exists an error is returned.
+	if _, err := s.FindByID(ctx, dbrp.OrganizationID, dbrp.ID); err == nil {
+		return ErrDBRPAlreadyExists("dbrp already exist for this particular ID. If you are trying an update use the right function .Update")
+	}
+	// If a dbrp with this orgID, db, and rp exists an error is returned.
+	if err := s.isDBRPUnique(ctx, *dbrp); err != nil {
+		return err
+	}
+
+	encodedID, err := dbrp.ID.Encode()
+	if err != nil {
+		return ErrInvalidDBRPID
+	}
+	b, err := json.Marshal(dbrp)
+	if err != nil {
+		return ErrInternalService(err)
+	}
+
+	return s.store.Update(ctx, func(tx kv.Tx) error {
+		bucket, err := tx.Bucket(bucket)
+		if err != nil {
+			return ErrInternalService(err)
+		}
+		if err := bucket.Put(encodedID, b); err != nil {
+			return ErrInternalService(err)
+		}
+		compKey := indexForeignKey(*dbrp)
+		if err := s.byOrgAndDatabase.Insert(tx, compKey, encodedID); err != nil {
+			return err
+		}
+		defSet, err := s.isDefaultSet(tx, compKey)
+		if err != nil {
+			return err
+		}
+		if !defSet {
+			dbrp.Default = true
+		}
+		if dbrp.Default {
+			if err := s.setAsDefault(tx, compKey, encodedID); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+// Updates a mapping.
+// If another mapping with same organization ID, database, and retention policy exists, an error is returned.
+// Un-setting `Default` for a mapping will cause the first one to become the default.
+func (s *Service) Update(ctx context.Context, dbrp *influxdb.DBRPMappingV2) error {
+	if err := dbrp.Validate(); err != nil {
+		return ErrInvalidDBRP(err)
+	}
+	oldDBRP, err := s.FindByID(ctx, dbrp.OrganizationID, dbrp.ID)
+	if err != nil {
+		return ErrDBRPNotFound
+	}
+	// Overwrite fields that cannot change.
+	dbrp.ID = oldDBRP.ID
+	dbrp.OrganizationID = oldDBRP.OrganizationID
+	dbrp.BucketID = oldDBRP.BucketID
+	dbrp.Database = oldDBRP.Database
+
+	// If a dbrp with this orgID, db, and rp exists an error is returned.
+	if err := s.isDBRPUnique(ctx, *dbrp); err != nil {
+		return err
+	}
+
+	encodedID, err := dbrp.ID.Encode()
+	if err != nil {
+		return ErrInternalService(err)
+	}
+	b, err := json.Marshal(dbrp)
+	if err != nil {
+		return ErrInternalService(err)
+	}
+
+	return s.store.Update(ctx, func(tx kv.Tx) error {
+		bucket, err := tx.Bucket(bucket)
+		if err != nil {
+			return ErrInternalService(err)
+		}
+		if err := bucket.Put(encodedID, b); err != nil {
+			return err
+		}
+		compKey := indexForeignKey(*dbrp)
+		if dbrp.Default {
+			err = s.setAsDefault(tx, compKey, encodedID)
+		} else if oldDBRP.Default {
+			// This means default was unset.
+			// Need to find a new default.
+			first, ferr := s.getFirstBut(tx, compKey, encodedID)
+			if ferr != nil {
+				return ferr
+			}
+			if len(first) > 0 {
+				err = s.setAsDefault(tx, compKey, first)
+			}
+			// If no first was found, then this will remain the default.
```

Comment:
Do we need to error here?
Since the caller is expecting to unset this as the default, they should probably be presented with something like:
`&influxdb.Error{Code: influxdb.Invalid, Msg: "cannot unset DBRP as default: cannot find alternative default"}`

Reply:
This is a design decision. Think this could be updated in a further PR if needed.